File: 3_2_qt_protocol_structure_demo/protocol/protocolmanager.cpp


#include "protocolmanager.h"
#include <QString>

namespace HYKT {

ProtocolManager *ProtocolManager::instance_p = nullptr;

ProtocolManager::ProtocolManager()
{
    is_sending_ = false;
    wait_response_ = false;
    pos = 0;
    package_on_ = false;


    //be note that caller-thread and creater-thread must be one thread for object of this class

    udp_manager_p = new NetManager();
    connect(udp_manager_p, &NetManager::onRecv, this, &ProtocolManager::onRecv);
    connect(udp_manager_p, &NetManager::onError, this, &ProtocolManager::onError);

    //for thread switch
    connect(this, &ProtocolManager::onDisconnect, this, &ProtocolManager::doDisconnect);
    connect(this, &ProtocolManager::callSend, this, &ProtocolManager::doSend);
    //connect(this, &ProtocolManager::callStartThread, this, &ProtocolManager::doStartThread);

    helper_thread_p = nullptr;
}

ProtocolManager::~ProtocolManager()
{
    disconnect(udp_manager_p, &NetManager::onRecv, this, &ProtocolManager::onRecv);
    disconnect(udp_manager_p, &NetManager::onError, this, &ProtocolManager::onError);

    disconnect(this, &ProtocolManager::onDisconnect, this, &ProtocolManager::doDisconnect);
    disconnect(this, &ProtocolManager::callSend, this, &ProtocolManager::doSend);
  //  disconnect(this, &ProtocolManager::callStartThread, this, &ProtocolManager::doStartThread);

    if(IsConnected()) {
        Disconnect();
    }

     udp_manager_p->deleteLater();

     if(helper_thread_p != nullptr)
         delete helper_thread_p;
}

ProtocolManager* ProtocolManager::GetInstance() {
    if (instance_p == nullptr) {
        instance_p = new ProtocolManager();
    }
    return instance_p;
}

/******************************* connection *****************************/

int ProtocolManager::Connect(QString ip, quint16 port) {
    return udp_manager_p->Open(Comm::NetType::UDP_client, ip, port);
}

void ProtocolManager::Disconnect() {
    udp_manager_p->Close();
}

bool ProtocolManager::IsConnected() {
    return udp_manager_p->IsOpen();
}

int ProtocolManager::Send(const QByteArray &data) {
    return udp_manager_p->Send(data);
}

void ProtocolManager::onRecv(QByteArray *data, QString ip, quint16 port) {
    Q_UNUSED(ip)
    Q_UNUSED(port)
    for(int a = 0; a < data->length(); a ++) {
        Process(static_cast<quint8>(data->at(a)));
    }
}

void ProtocolManager::onError(int error_code, QString extra_error_info) {
    emit onConnectionError(QString::number(error_code)+ extra_error_info);
}

void ProtocolManager::doDisconnect() {
    Disconnect();
}

/******************************* thread *****************************/

void ProtocolManager::StartThread() {
    is_sending_ = true;
    start();
}

void ProtocolManager::StopThread() {
    is_sending_ = false;
}

bool ProtocolManager::IsThreadWorking() {
    return is_sending_;
}

void ProtocolManager::run() {
    while(is_sending_) {
        current_work_p = wm_.GetCurrentWork();
        if (current_work_p != nullptr) {
            wait_response_ = true;
            emit callSend();

            start_work_secs_ = QDateTime::currentSecsSinceEpoch();

        } else { //end of all work
            StopThread();
            wait_response_ = false;
        }

        while (wait_response_) {
             msleep(10);
             //call re-send when timeout
            qint64 current_secs = QDateTime::currentSecsSinceEpoch();
            if(current_secs - start_work_secs_ > TIME_OUT_SECONDS) {
                ReSendPkg();
            }

        }
    }
}

void ProtocolManager::doSend() {
    Send(current_work_p->data);
}

/********************** work process **************************/

void ProtocolManager::AddWork(QByteArray &pkg) {
    wm_.AddWork(pkg);
     if (IsThreadWorking() == false) {
         StartThread();
     }
}

void ProtocolManager::CleanWorkList() {
    wm_.CleanWorkList();
}

void ProtocolManager::Process(quint8 b) {
    if (package_on_ == true)
    {
        if (pos > CACHE_LENGTH) {
            package_on_ = false;
            pos = 0;
            return;
        }

        cache[pos++] = b;

        if (pos > 4 &&
            cache[pos - 4] == tail[0] &&    //a pkg is finished
            cache[pos - 3] == tail[1] &&
            cache[pos - 2] == tail[2] &&
            cache[pos - 1] == tail[3] &&
            cache[sizeof (head)] + sizeof (head) + sizeof (tail)  == pos //check pkg len
                ) {
            package_on_ = false;

            int skip_head_and_pkg_len = 2;
            int skip_tail = 4;
            quint8* pkg = &cache[skip_head_and_pkg_len]; //remove head and pkg-len
            int pkg_len = pos - skip_head_and_pkg_len - skip_tail;//skip index of tail byte

            if (Crc::CheckCRC(pkg, pkg_len) == true) {
                ProcessMetaPkg(pkg, pkg_len-2);//-2 : remove crc byte
            } else {
                CleanWorkList();
                emit onConnectionError("crc check error");
            }
        }
    } else if (package_on_ == false && b == head[0]) { //start a pkg
        package_on_ = true;
        pos = 0;
        cache[pos++] = b;
    }
}

//for timer send a false pkg
void ProtocolManager::ReSendPkg() {
     ProcessMetaPkg(nullptr, 0);
}

//thread and recv-callback are will call this functinn ,so this function need a lock
void ProtocolManager::ProcessMetaPkg(quint8 *pkg, int len) {
    mutex_.lock();

    //根据指令类型码判断，如果是下位机主动反馈指令，在这里拦截判断，else 再执行以下判断
    if(Command::IsCommandIsReportingType(pkg)) {
        DispatchPackage(pkg, len);
        mutex_.unlock();
        return;
    }

    if (wait_response_ == false) {
        mutex_.unlock();
        return;
    }

    //检验下位机对上位机命令的反馈指令是否正确
    if (pkg != nullptr && current_work_p->CheckResponse(pkg) == true) {
        wm_.FinishAWork();
        DispatchPackage(pkg, len);
    } else {
        //mark timeout
        if (current_work_p != nullptr)
        {
            current_work_p->failed_times++;
            if (current_work_p->failed_times >= 3)
            {
                CleanWorkList();
                StopThread();
                wait_response_ = false;
               // Disconnect();
                emit onDisconnect();

                emit onConnectionError("after re-try 3 times, connection broken");
            }
        }
    }
    wait_response_ = false;

    mutex_.unlock();
}

//对于接收下位反馈给上位机数据的指令，在此函数中获取数据
void ProtocolManager::DispatchPackage(quint8 *pkg, int len) {
    if (len < 2) return;

    if(Command::IsCommandReportValue(pkg))  {
        Send(*Command::ResponseForReportValue()); //give response immediately
        emit onReportValue(Command::GetReportValue(pkg, len));
        //wm_.InsertWorkAtHead(*Command::ResponseForReportValue());
    }

    //根据不同指令，调用对应的处理函数，并调用其回调函数，如果有必要的时候
    if(Command::IsCommandGetMachinetemperature(pkg)) {
        emit onGetMachinetemperature(Command::GetValueOfGetMachinetemperature(pkg, len));
    }
}


} //namespace HYKT

File: 3_2_winform_protocol_structure_demo/3_2_winform_protocol_structure_demo/Protocol/ProtocolManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading;
using System.Timers;


namespace HYKT
{
    class ProtocolManager
    {
        private const string TAG = "ProtocolManager";
        private NetManager netManater;
        private static WorkManager wm;
        private static System.Timers.Timer aTimer = null;
        private const int DataTranferTimeoutTime = 4000;


        private Thread mythread;
        private static bool isRun = false;
        private static bool waitResponse = false;
        private Work work = null;

        private bool package_on = false;
        private byte[] cache;
        private int pos = 0;
        private static readonly int CACHE_LENGTH = 1024 * 5;

        private static readonly int ErrorCodeNetError = -1;
        private static readonly int ErrorCodeFailed3TimesError = -2;
        private static readonly int ErrorCodeCrcError = -3;

        public delegate void OnProtocolRecvListener(byte[] pkg);
        public delegate void OnProtocolErrorListener(int code, string reason);
        public event OnProtocolRecvListener eventProtocolRecvHandler;
        public event OnProtocolErrorListener eventProtocolErrorHandler;

        private void CallProtocolRecv(byte[] pkg) {
            if(eventProtocolRecvHandler != null)
            {
                eventProtocolRecvHandler.Invoke(pkg);
            }
        }

        private void CallProtocolError(int code, string reason)
        {
            if (eventProtocolErrorHandler != null)
            {
                eventProtocolErrorHandler.Invoke(code, reason);
            }
        }

        public delegate void onGetMachinetemperature(int value);
        public delegate void onReportValue(int value);

        private onGetMachinetemperature getMachinetemperatureListener = null;
        private onReportValue onReportValueListener = null;

        public void SetOnGetMachineTemperatureListener(onGetMachinetemperature l ) { getMachinetemperatureListener = l; }
        public void SetOnReportValueListener(onReportValue l) { onReportValueListener = l; }



        private ProtocolManager() {
            Init();
        }

        private static ProtocolManager instance = null;
        public static ProtocolManager GetInstance() {
            if (instance == null) {
                instance = new ProtocolManager();
            }
            return instance;
        }

        public static void CleanInstance()
        {
            if (wm != null) {
                wm.CleanWorkList();
            }
        }

        private void Init() {
            cache = new byte[CACHE_LENGTH];

            wm = new WorkManager();
            netManater = new NetManager();
            netManater.SetOnNetRecvListener(OnNetRecvListener);
            netManater.SetOnNetErrorListener(OnNetErrorListener);

            StartWorkThread();
        }

        /*********************************** work manager **********************************/
        public void AddWork(byte[] data)
        {
            wm.AddWork(data);
            if (IsThreadWorking() == false)
            {
                StartWorkThread();
            }
        }

        public void CleanWorkList()
        {
            wm.CleanWorkList();
        }
        public bool IsHasWork()
        {
            return wm.IsHasWork();
        }

        /********************************** timer *********************************/
        private void StartTimer() {
            if (aTimer != null && aTimer.Enabled == false)
            {
                aTimer.Enabled = true;
            }
            else
            {
                aTimer = new System.Timers.Timer(DataTranferTimeoutTime);
                aTimer.Elapsed += new ElapsedEventHandler(OnTimedEvent);
                aTimer.AutoReset = false;
            }
        }

        private void StopTimer() {
            aTimer.Enabled = false;
        }

        private void OnTimedEvent(object source, ElapsedEventArgs e)
        {
            ProcessResult(null);//send a empty pkg
        }

        /***************************** thread **************************************/
        private void StartWorkThread() {
            mythread = null;
            mythread = new Thread(ThreadWork);
            isRun = true;
            mythread.Start();
        }

        private static void StopWorkThread() {
            waitResponse = false;
            isRun = false;
        }

        private bool IsThreadWorking() {
            return isRun;
        }

        private void ThreadWork()
        {
            while (isRun) {
                work = wm.GetWork();
                if (work != null)
                {
                    Send(work.Data);
                    waitResponse = true;
                }
                else { //end of all work
                    StopWorkThread();
                }

                while (waitResponse) {
                    Thread.Sleep(1);
                }
            }
        }

        /******************************** connection functions *****************************/
        public bool Connect(string addr, int port) {
            return netManater.Open(CommNetType.NetType.UDP_client, addr, port);
        }

        public void DisConnect() {
            netManater.Close();
        }

        public bool IsConnected()
        {
            return netManater.IsOpen();
        }

        public void Send(byte[] data)
        {
            netManater.Send(data);
            StartTimer();//set a timer for response timeout
        }

        private void OnNetRecvListener(byte[] data, int length, IPEndPoint iped)
        {
            for (int a = 0; a < length; a++)
            {
                Process(data[a]);
            }
        }

        private void OnNetErrorListener(string reasonStr, Exception e)
        {
            CallProtocolError(ErrorCodeNetError, reasonStr);
        }

        private void Process(byte b) {
            if (package_on == true)
            {
                if (pos > CACHE_LENGTH) {
                    package_on = false;
                    pos = 0;
                    return;
                }

                cache[pos++] = b;
                if (pos > 4 &&
                    cache[pos - 4] == Command.tail[0] &&    //a pkg is finished
                    cache[pos - 3] == Command.tail[1] &&
                    cache[pos - 2] == Command.tail[2] &&
                    cache[pos - 1] == Command.tail[3] &&
                    cache[Command.head.Length] + Command.head.Length + Command.tail.Length == pos //check pkg len
                    ) {
                    package_on = false;

                    byte[] temp = cache.Take(pos - 4).ToArray();//remove tail' bytes
                    byte[] pkg = temp.Skip(2).ToArray();//remove head byte and pkg-len byte

                    if (CRC16.CheckCRC(pkg) == true)
                    {
                        ProcessResult(pkg.Take(pkg.Length - 2).ToArray());//remove crc's bytes
                    }
                    else
                    {
                        CleanWorkList();
                        CallProtocolError(ErrorCodeCrcError, "crc check error");
                    }
                }
            }
            else if (package_on == false && b == Command.head[0]) { //start a pkg
                package_on = true;
                pos = 0;
                cache[pos++] = b;
            }
        }


        static object _object = new object();// for locker
        private void ProcessResult(byte[] pkg)
        {
            lock (_object)
            {
                //根据指令类型码判断，如果是下位机主动反馈指令，在这里拦截判断，else 再执行以下判断
                if (Command.IsCommandIsReportingType(pkg))
                {
                    DispatchPackage(pkg);
                    return;
                }

                if (waitResponse == false) return;

                //检验下位机对上位机命令的反馈指令是否正确
                if (pkg != null && work.CheckResponse(pkg) == true)
                {
                    StopTimer();
                    wm.FinishAWork();
                    DispatchPackage(pkg);
                    CallProtocolRecv(pkg);
                    waitResponse = false;
                }
                else
                {
                    StopTimer();
                    //mark timeout
                    if (work != null)
                    {
                        work.FailedTimes++;
                        if (work.FailedTimes >= 3)
                        {
                            CleanWorkList();
                            netManater.Close();
                            CallProtocolError(ErrorCodeFailed3TimesError, "re-send 3 times error");
                        }
                    }
                    waitResponse = false;
                }
            }
        }

        //对于接收下位反馈给上位机数据的指令，在此函数中获取数据
        private void DispatchPackage(byte[] pkg)
        {
            if (pkg.Length < 2) return;

            if (Command.IsCommandReportValue(pkg))
            {
                netManater.Send(Command.ResponseForReportValue());//give response immediately
                if (onReportValueListener != null) {
                    onReportValueListener(Command.GetReportValue(pkg));
                }
            }

            //根据不同指令，调用对应的处理函数，并调用其回调函数，如果有必要的时候
            if (Command.IsCommandGetMachinetemperature(pkg))
            {
                if (getMachinetemperatureListener != null) {
                    getMachinetemperatureListener(Command.GetValueOfGetMachinetemperature(pkg));
                }
            }
        }


    }
}
